src/InstrumentationEngine.Attach/Program.cs

```diff
@@ -0,0 +1,170 @@
+                return;
+            }
+
+            string? rootDirectory = Path.GetDirectoryName(toolsDirectory);
+            if (null == rootDirectory || !Directory.Exists(rootDirectory))
+            {
+                WriteErrorJson(Invariant($"Directory '{rootDirectory}' does not exist."));
+                return;
+            }
+
+            string enginePath;
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                enginePath = Path.Combine(rootDirectory, "Instrumentation64", "libInstrumentationEngine.so");
+            }
+            else
+            {
+                bool isWow64;
+                try
+                {
+                    // isWow64 is true for 32-bit applications running on a 64-bit OS
+                    if (!NativeMethods.IsWow64Process(process.Handle, out isWow64))
+                    {
+                        Exception? ex = Marshal.GetExceptionForHR(Marshal.GetHRForLastWin32Error());
+                        if (null != ex)
+                        {
+                            WriteErrorJson(Invariant($"Failed to check process bitness: {ex.Message}"));
+                        }
+                        else
+                        {
+                            WriteErrorJson(Invariant($"Failed to check process bitness."));
+                        }
+                    }
+                }
+                catch (Win32Exception ex)
+                {
+                    WriteErrorJson(Invariant($"Failed to check process bitness (code: {ex.NativeErrorCode}): {ex.Message}"));
+                    return;
+                }
+
+                enginePath = isWow64 ?
+                    Path.Combine(rootDirectory, "Instrumentation32", "MicrosoftInstrumentationEngine_x86.dll") :
+                    Path.Combine(rootDirectory, "Instrumentation64", "MicrosoftInstrumentationEngine_x64.dll");
+            }
+
+            if (!File.Exists(enginePath))
+            {
+                WriteErrorJson(Invariant($"Engine path '{enginePath}' does not exist."));
+                return;
+            }
+
+            // CONSIDER: Should the engine be validated before attempting to attach it to the CLR?
+            // For example, could check the signature on the assembly (which wouldn't work on Linux).
+
+            DiagnosticsClient client = new DiagnosticsClient(processId);
+            try
+            {
+                client.AttachProfiler(TimeSpan.FromSeconds(10), InstrumentationEngineClsid, enginePath);
+            }
+            catch (ServerErrorException ex)
+            {
+                WriteErrorJson(Invariant($"Could not attach engine to process: '{ex.Message}'."));
+                return;
+            }
+
+            WriteSuccessJson();
+        }
+
+        private static void WriteErrorJson(string message)
+        {
+            Console.WriteLine(CreateResultJson("Failed", message));
```

Comment:
>CreateResultJson [](start = 30, length = 16)

Why in Json?

Reply:
The executable is going to be executed by an http handler. I felt it was best to use a standard data format that it could easily consume.

---
In reply to: [384719780](https://github.com/microsoft/CLRInstrumentationEngine/pull/212#discussion_r384719780) [](ancestors = 384719780)

Reply:
Maybe I should have it write messages to standard error and return a non-zero exit code. Then let the http handler convert it to whatever it needs.

---
In reply to: [384732346](https://github.com/microsoft/CLRInstrumentationEngine/pull/212#discussion_r384732346) [](ancestors = 384732346,384719780)